### PR TITLE
update the inference framework to support new ARD kernels

### PR DESCRIPTION
**Pull request by yorkerlin**
@karlnapf 
take a look at this
This PR is about clean-up and updating the framework.
Because  parameters of all kernels (expect the new ARD kernels) are either scalar or vector, the GP framework without this change only work for computing the gradient wrt scalar and vector.
This PR improves the framework to support parameters of a kernel in a matrix form.

### CODE REVIEW

File: src/shogun/machine/gp/EPInferenceMethod.cpp [hunk 1]
```diff
@@ -64,6 +64,18 @@ void CEPInferenceMethod::init()
 	m_tol=1e-4;
 }
 
+CEPInferenceMethod* CEPInferenceMethod::obtain_from_generic(
+		CInferenceMethod* inference)
+{
+	ASSERT(inference!=NULL);
```
Comment (karlnapf) [inline src/shogun/machine/gp/EPInferenceMethod.cpp]: REQUIRE

File: src/shogun/machine/gp/ExactInferenceMethod.cpp [hunk 2]
```diff
@@ -270,6 +270,18 @@ SGVector<float64_t> CExactInferenceMethod::get_derivative_wrt_inference_method(
 	return result;
 }
 
+CExactInferenceMethod* CExactInferenceMethod::obtain_from_generic(
+		CInferenceMethod* inference)
+{
+	ASSERT(inference!=NULL);
```
Comment (karlnapf) [inline src/shogun/machine/gp/ExactInferenceMethod.cpp]: REQUIRE

File: src/shogun/machine/gp/KLApproxDiagonalInferenceMethod.cpp [hunk 3]
```diff
@@ -71,6 +71,18 @@ void CKLApproxDiagonalInferenceMethod::init()
 		MS_NOT_AVAILABLE);
 }
 
+CKLApproxDiagonalInferenceMethod* CKLApproxDiagonalInferenceMethod::obtain_from_generic(
+		CInferenceMethod* inference)
+{
+	ASSERT(inference!=NULL);
```
Comment (karlnapf) [inline src/shogun/machine/gp/KLApproxDiagonalInferenceMethod.cpp]: REQUIRE

File: src/shogun/machine/gp/SingleLaplacianInferenceMethod.cpp [hunk 4]
```diff
@@ -166,7 +178,29 @@ void CSingleLaplacianInferenceMethod::update_approx_cov()
...
+	m_W=m_d2lp.clone();
+	m_W.scale(-1.0);
+	m_sW=SGVector<float64_t>(m_W.vlen);
+
+	// compute sW
 	Map<VectorXd> eigen_W(m_W.vector, m_W.vlen);
+	Map<VectorXd> eigen_sW(m_sW.vector, m_sW.vlen);
+
+	if (eigen_W.minCoeff()>0)
+	{
```
Comment (karlnapf) [inline src/shogun/machine/gp/SingleLaplacianInferenceMethod.cpp]: no {}

File: src/shogun/machine/gp/SingleLaplacianInferenceMethod.cpp [hunk 5]
```diff
@@ -177,19 +211,18 @@ void CSingleLaplacianInferenceMethod::update_chol()
...
 		FullPivLU<MatrixXd> lu(
-			eigen_ktrtr*CMath::sq(m_scale)+MatrixXd(eigen_iW.asDiagonal()));
-
-		// compute cholesky: L = -(K + iW)^-1
-		eigen_L = -lu.inverse();
+			MatrixXd::Identity(m_ktrtr.num_rows,m_ktrtr.num_cols)+
+			eigen_ktrtr*CMath::sq(m_scale)*eigen_W.asDiagonal());
+		// compute cholesky: L = -(K + 1/W)^-1
+		//-iA = -inv(A)
+		eigen_L=-lu.inverse();
```
Comment (karlnapf) [inline src/shogun/machine/gp/SingleLaplacianInferenceMethod.cpp]: hopefully soon via linalg?
@lambday how are the linear solves going?

File: src/shogun/machine/gp/SingleLaplacianInferenceMethod.h [hunk 6]
```diff
@@ -111,7 +124,11 @@ class CSingleLaplacianInferenceMethod: public CLaplacianInferenceBase
 	 */
 	virtual SGVector<float64_t> get_diagonal_vector();
 
+	virtual void update();
```
Comment (karlnapf) [inline src/shogun/machine/gp/SingleLaplacianInferenceMethod.h]: please document

**Comment by karlnapf** [comment]
looks good to me, a few minor style issues.

Is this covered by unit tests?

**Comment by yorkerlin** [comment]
This change is  tested in `FITCInferenceMethod`.

**Comment by yorkerlin** [comment]
@karlnapf 
do you know why this PR automatically closed when I `git push origin +GPInference`

**Comment by yorkerlin** [comment]
@karlnapf  Please take a look at
that PR first https://github.com/shogun-toolbox/shogun/pull/2736
I will resend a new PR since I cannot reopen this PR.

**Comment by karlnapf** [comment]
Dont know why, but will look at the other

**Pull request closed without merging**